tc-messaging/src/main/java/com/tc/io/TCByteBufferInputStream.java

```diff
@@ -273,74 +164,68 @@ public final TCByteBuffer read(int len) {
     if (available() == 0) { return TCByteBufferFactory.getInstance(0); }
 
     TCByteBuffer result = TCByteBufferFactory.getInstance(len);
-    while (this.index < this.numBufs) {
-      TCByteBuffer buf = this.data[this.index];
-      if (buf.hasRemaining()) {
-        int limit = buf.limit();
-        buf.limit(Math.min(buf.position() + result.remaining(), limit));
-        result.put(buf);
-        buf.limit(limit);
-      }
-      if (!result.hasRemaining()) {
-        break;
-      } else {
-        nextBuffer();
-      }
+    TCByteBuffer src = nextBuffer();
+    while (result.hasRemaining()) {
+      int limit = src.limit();
+      src.limit(Math.min(src.position() + result.remaining(), limit));
+      result.put(src);
+      src.limit(limit);
+      src = nextBuffer();
     }
     if (result.hasRemaining()) {
       throw new BufferUnderflowException();
     }
     return result.flip();
   }
-
+  
   @Override
-  public final int read() {
+  public final TCReference readReference(final int len) {
     checkClosed();
 
-    while (this.index < this.numBufs) {
-      if (this.data[this.index].hasRemaining()) {
-        this.position++;
-        return this.data[this.index].get() & 0xFF;
+    if (len == 0) { 
+      return TCReferenceSupport.createGCReference(Collections.singleton(TCByteBufferFactory.getInstance(0))); 
+    }
+    
+    TCReference dup = data.duplicate();
```

Comment:
This code sitting outside of TCReference feels a bit wrong. Could we introduce say a `TCReference.duplicate(int length)` call that duplicates and truncates in a single call? That way we wouldn't have external code manipulating a member variable of the TCReference?

Reply:
I can see that.  Feels weird manipulating the contents outside.  My justification for doing it this was to not to overload TCReference with too much functionality.  I wanted it to just be a container for ByteByffers.  Part of the problem is that a bunch of this stuff gets manipulated outside at all the layers.  TCByteBuffer.position and limit for example.  Even worse there is nothing preventing someone from holding on the the byte buffer and closing the reference.  Really, it would be nice to have just the reference and no bytebuffers but we are a long way from that.